C++: Add experimental Array Length Tracking library

Cornelius-Riemenschneider opened the pull request:
This PR starts the implementation of  a new library in experimental that tracks the length of the memory a pointer points to, as well as the offset of that pointer in the chunk of memory.

For the design outline, see the header comment in ArrayLengthAnalysis.qll.
The design goal for this library is to be always sound, when the analyzed program is not exhibiting undefined behvavior.

Currently, only C-ish constructs are supported and tested.
I began working on C++ constructs (like this pointer, and reference types), but in order to keep the PR review-able size-wise, these features have been omitted now.

The big picture:
Eventually, I want to make this library intraprocedural.
Furthermore, as a test case for this library, I will also introduce a test that checks all pointer dereferences in a program and attempts to prove that they are in-bounds.
Implementing these steps piece-wise makes the code review hopefully more pleasant.


This PR depends on #3337

MathiasVP commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+
+  VNLength() { this = TVNLength(vn) }
+
+  /** Gets the SSA variable that equals value number bound. */
```
```suggestion
  /** Gets an instruction with this value number bound. */
```
might be more appropriate description?

Cornelius-Riemenschneider replied:
True, I've used your suggestion. However, my text matches Bound.qll, line 71, so if you feel inclined that wording over there should also be adjusted.

MathiasVP replied:
Good point!

jbj commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+ * All pointers in this predicate are guaranteed to be non-null,
+ * but are not guaranteed to be live.
+ */
+cached
```
I recommend avoiding `cached` until you've tested that it works as intended in queries. It can be hard to get it right. For example, should the newtypes for `Length` and `Offset` be cached? It's hard to know without measuring.

jbj commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+ * Holds if the instruction `array` represents a pointer to a chunk of memory that holds
+ * `length + lengthDelta` elements, using only local analysis.
+ * `array` points at `offset + offsetDelta` in the chunk of memory.
+ * The pointer is in-bounds if `offset < length + lengthDelta` and
+ * `offset + offsetDelta >= 0` holds.
```
Shouldn't you add `offsetDelta` to `offset` in all cases? So the constraint is `0 <= offset + offsetDelta < length + lengthDelta`. Same in the line below.

jbj commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+ * This class keeps track of the ValueNumber or Zero.
+ * The delta is tracked in the predicate `knownArrayLength`.
+ */
+abstract class Length extends TLength {
+  abstract string toString();
```
Users of your library will generally have an easier time if you don't use `abstract`.

```suggestion
class Length extends TLength {
  string toString() { none() } // overridden in subclasses
```

Cornelius-Riemenschneider replied:
I don't have an opinion on this, so I'm following your suggestion.

jbj commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+ * It is either an `Operand` or zero, an additional integer delta is added later.
+ */
+abstract class Offset extends TOffset {
+  abstract string toString();
```
Same here, about `abstract`.

jbj commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+/**
```
The directory structure under `experimental` should mirror the structure of `src`, so I'd like to move this library to a directory that mirrors it future location outside `experimental`. I think it could belong in `semmle/code/cpp/rangeanalysis`. What do you think, @rdmarsh2?

rdmarsh2 replied:
That sounds right.

Cornelius-Riemenschneider commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+
+private newtype TLength =
+  TZeroLength() or
+  TVNLength(ValueNumber vn) {
+    exists(Instruction i |
+      vn.getAnInstruction() = i and
+      (
+        i.getResultIRType() instanceof IRSignedIntegerType or
+        i.getResultIRType() instanceof IRUnsignedIntegerType
+      )
+    ) and
+    not vn.getAnInstruction() instanceof ConstantInstruction
```
Actually, the way I think about the length of the array here is that often, to get useful information about the array length, the user will invoke the Range Analysis library on the length.
Thus, it makes sense to copy over the list of instruction types from there.
Then there's still the difference that `Bound` restructs the types to IntegralType and PointerType, whereas a `Length` only makes sense for an integer type (and, pending #2217, the code in `Bounds.qll` needs to be adjusted).

rdmarsh2 replied:
You can also drop the `FieldAddressInstruction` and `VariableAddressInstruction` cases, since their results should always be pointers.

Cornelius-Riemenschneider commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+/**
+ * Parses a `sizeExpr` of malloc into a variable part (`lengthExpr`) and an integer offset (`delta`).
+ */
+private predicate deconstructMallocSizeExpr(Expr sizeExpr, Expr lengthExpr, int delta) {
```
We don't get around mapping between AST and IR, because I still want to use Allocation.qll/getSizeMult() and getSizeExpr().

In the very first iteration of this I had the whole size-logic IR-based, but it wasn't really a fluent way to write it.
Especially sizeof is gone at the IR level, so we need the AST to figure out the size multiplicator anyways.
If we have to do the conversion work anyways, then I don't see why we should move the pattern matching against size expressions from the AST to the IR.
On the other hand, using `RangeUtils.getConstantValue()`, it is quite easy so if you have a preference of IR over AST I can certainly follow through.
What do you think?

rdmarsh2 replied:
The advantage I see for using the IR when calculating the size would be handling something like the following:
```
size_t len= count* sizeof(struct s);
struct s *arr = malloc(len);
memset(arr, 0, len);
```
That shouldn't be an obstacle for going into experimental, though.

Cornelius-Riemenschneider replied:
I don't see how implementing the size-matching in the IR would help with that, as the IR wouldn't inline the definition of len, would it?
Anyways, I haven't thought about this pattern of allocating something with count*sizeof() and then passing the length in bytes somewhere.

As this is a more complicated problem than the one I set out to solve, I agree that we can defer this for now.

Cornelius-Riemenschneider commented on cpp/ql/src/experimental/library/ArrayLengthAnalysis.qll:
```diff
@@ -0,0 +1,274 @@
+      (
+        exists(Expr lengthExpr |
+          deconstructMallocSizeExpr(alloc.getSizeExpr(), lengthExpr, delta) and
+          // TODO converted or unconverted here?
```
Thanks!

Cornelius-Riemenschneider replied:
I'm not sure this is the results we want.
See https://github.com/github/codeql/commit/f83c3452a1c17b42ae12b105110429d4ca1d91c4
for the diff in test output, we now track both the unconverted and the converted length.
However, if we use the range analysis to get information about the array length, tracking the converted length does not buy us anything, as the range analysis ignores upcasts anyways.

rdmarsh2 replied:
oh, I misremembered what `getUnconvertedResultExpression` did. `getConvertedResultExpression` is the right one here.

Cornelius-Riemenschneider commented:
Thank you for your reviews!
I addressed all simple points, and replied to the more complicated ones.
For moving the library, I'm waiting on @rdmarsh2 's opinion.

About `simpleArrayLengthStep`:
I don't think that the Cartesian product actually occurs.
Jonas told me that, due to `bindingset[prevOffsetDelta]`, the predicate gets inlined, and prevOffset is always bound.
The motivation to have this as a separate predicate is only to increase the readability of `knownArrayLength`, in fact it should be marked `private`.
I see these options forward:
1) add prevOffset to the bindingSet to make clear that it is an input  parameter
2) inline the predicate by hand
3) do the refactoring suggested by Robert, and investigate how to deal with the change of offset in the one case where it does change
With my understanding of the QL evaluation model, I lean towards option 1, as it stays with the motivation (factor out the code as a separate predicate to increase readability) the closest.

rdmarsh2 commented:
Right, that forced inlining will bind `prevOffset`. The first option is easiest, and probably the right approach for the moment.

In the range analysis library, splitting the step relation into its own set of predicates that don't depend on the main recursion was a performance optimization. I suspect it would be faster for this as well, but is not worth doing at this point.

Cornelius-Riemenschneider commented:
I addressed all review comments.

Splitting the step relation out into a separate predicate is probably worthwhile investigating when tuning the performance of this on large projects.
So far, I've run this on C projects up until 100k LOC and it was quite fast, usually the IR construction time dominated the execution time.

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeUtils.qll:
```diff
@@ -80,3 +80,44 @@ predicate backEdge(PhiInstruction phi, PhiInputOperand op) {
+  }
+}
+
+class SafeIntCastInstruction extends ConvertInstruction {
```
```suggestion
/**
 * A `ConvertInstruction` which casts from one integer type to another in a way
 * that cannot overflow or underflow.
 */
class SafeIntCastInstruction extends ConvertInstruction {
```

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeUtils.qll:
```diff
@@ -80,3 +80,44 @@ predicate backEdge(PhiInstruction phi, PhiInputOperand op) {
+  SafeIntCastInstruction() { safeCast(getUnary().getResultType(), getResultType()) }
+}
+
+class SafeCastInstruction extends ConvertInstruction {
```
```suggestion
/**
 * A `ConvertInstruction` which does not invalidate bounds determined by
 * range analysis.
 */
class SafeCastInstruction extends ConvertInstruction {
```

Cornelius-Riemenschneider commented:
Good to hear! I addressed your comments.
